### ISSUE TITLE
Deterministic salts

### ISSUE THREAD

**Issue by JeremyRand** [namecoin/meta#58]
Background: https://github.com/namecoin/electrum-nmc/issues/237

Namecoin wallets (Namecoin Core and Electrum-NMC) currently use a random salt for name registrations.  While this is relatively easy to reason about security-wise, it means that salts cannot be recovered from an HD seed, i.e. if a user restores a wallet from a seed while halfway through a name registration, the name registration cannot be completed.

I asked @ryancdotorg on IRC how we could deterministically derive salts.  He said that the standard crypto function for this is [HKDF](https://tools.ietf.org/html/rfc5869).  Specifically, we'd use it with the following parameters:

* IKM: private key of the address owning the `name_new`.  (I'm not certain how we should handle multisig addresses, maybe just pick whichever private key belongs to the cosigner who creates the transaction?)
* Salt: name identifier.
* Info: `Namecoin Registration Salt`
* L: `20` (max length of the Namecoin salt)

The standard Python implementation is at https://github.com/casebeer/python-hkdf ; it's under the BSD 2-clause license, so should be fine for Electrum-NMC.  I'm not sure what C/C++ implementations would be appropriate for Namecoin Core (would need to be MIT-compatible).

**Comment by JeremyRand** [comment]
@domob1812 Do you have any comments on this?  Ideally Electrum-NMC and Namecoin Core would use the same logic for this, although it's not strictly necessary to do so in order to get some tangible benefit.

**Comment by domob1812** [comment]
That makes sense to me, although of course you would need to implement the corresponding wallet logic as well for recovering such registrations and filling in the "prepared `name_firstupdate`" when restoring a backup from seed.  (But that can always be done later as a new feature or manually if needed.)

Although I have to say it seems fairly low priority to me; this is something that will happen only very rarely, and even if it does happen, then the user has essentially just lost 0.01 NMC and can redo the registration with the new wallet.

I have no objections to implementing this in Namecoin Core, though.  What if instead of HKDF (and requiring a new dependency) we just SHA-256 (or something else readily implemented in the codebase) hashed the private key and name identifier together?  Assuming there is no preimage weakness for the hash function, this should be as secure?

**Comment by JeremyRand** [comment]
> Although I have to say it seems fairly low priority to me; this is something that will happen only very rarely, and even if it does happen, then the user has essentially just lost 0.01 NMC and can redo the registration with the new wallet.

@domob1812 I don't think it's critical to do immediately, but would be nice to have.

>  What if instead of HKDF (and requiring a new dependency) we just SHA-256 (or something else readily implemented in the codebase) hashed the private key and name identifier together? Assuming there is no preimage weakness for the hash function, this should be as secure?

I don't think we can be confident that it will be as secure.  For example, there might be cross-protocol attacks if we don't use an info string.

Anyway it looks like Bitcoin Core already has an HKDF implementation.  https://github.com/bitcoin/bitcoin/blob/99813a9745fe10a58bedd7a4cb721faf14f907a4/src/crypto/hkdf_sha256_32.cpp

Its length is hardcoded to 32 bytes whereas we want 20 bytes, but I suspect that's an easy patch.

**Comment by domob1812** [comment]
Good catch, we can certainly use that.  How exactly does the output size play into this?  SHA-256 by itself has 32 bytes outputs, so would we then just crop it to the first 20 bytes?  Or is the process more complex than that?

**Comment by ryancdotorg** [comment]
Yes, you can just truncate.

**Comment by JeremyRand** [comment]
> The standard Python implementation is at https://github.com/casebeer/python-hkdf

Update: both `Cryptodome` and `cryptography`, which Electrum already uses for crypto functions, include an HKDF implementation.  So there's no need to add the `python-hkdf` library as a dependency.
